astrokat/observe_main.py

```diff
@@ -537,10 +561,34 @@ def run_observation(opts, kat):
                         )
                     )
                     continue
+
             # TODO: setup of noise diode pattern should be moved to sessions
             #  so it happens in the line above
-            if "noise_diode" in obs_plan_params.keys():
-                noisediode.pattern(kat.array, session, obs_plan_params["noise_diode"])
+            if "noise_diode" in obs_plan_params:
+                nd_setup = obs_plan_params["noise_diode"]
+                nd_lead = None
+                if "lead_time" in nd_setup:
+                    nd_lead = nd_setup['lead_time']
+
+                # Set noise diode period to multiple of correlator integration time.
+                if not kat.array.dry_run:
+                    cbf_corr = session.cbf.correlator
+                    dump_period = cbf_corr.sensor.int_time.get_value()
+                else:
+                    dump_period = 0.5  # sec
+                user_logger.debug('DEBUG: Correlator integration time {} [sec]'
+                                  .format(dump_period))
+
+                if "cycle_len" in nd_setup:
+                    cycle_len_frac = -(nd_setup['cycle_len'] // -dump_period)
+                    nd_setup['cycle_len'] = cycle_len_frac * dump_period
+                    msg = 'Set noise diode period to multiple of correlator dump period:'
+                    msg += ' cycle length = {} [sec]'.format(nd_setup['cycle_len'])
+                    user_logger.warning(msg)
+                    noisediode.pattern(kat.array,
+                                       nd_setup,
+                                       lead_time=nd_lead,
+                                       )
```

Comment:
Implementation and requirement for this is unknown
Cannot comment on future undefined commissioning requirements

These will have to be provided by the astronomers with usage examples

Reply:
OK, I must have missed something then.

Reply:
@mserylak Good question - I tried the new and old algorithms (from previous comment) with cycles times shorter than a dump period.  The new method seems have a lower bound of one dump period, which sound good.  Old method would have cycle time of zero, which would break things.

Does it make sense to have a shorter cycle time than the dump period?

```python
In [4]: for c, d in zip([0.4, 0.5, 0.6, 0.4, 0.5, 0.6], [0.5, 0.5, 0.5, 0.499975, 0.499975, 0.499975]): 
   ...:     print('cycle len {}, dump period {} => old: {}, new: {}'.format(c, d, old(c, d), new(c, d))) 
   ...:                                                                                                                                                                    
cycle len 0.4, dump period 0.5 => old: 0.0, new: 0.5
cycle len 0.5, dump period 0.5 => old: 0.5, new: 0.5
cycle len 0.6, dump period 0.5 => old: 0.5, new: 1.0
cycle len 0.4, dump period 0.499975 => old: 0.0, new: 0.499975
cycle len 0.5, dump period 0.499975 => old: 0.499975, new: 0.99995
cycle len 0.6, dump period 0.499975 => old: 0.499975, new: 0.99995
```